lib/chat_channel_archive_service.rb

```diff
@@ -0,0 +1,200 @@
+    return if ChatChannelArchive.exists?(chat_channel: chat_channel)
+
+    ChatChannelArchive.transaction do
+      chat_channel.read_only!(acting_user)
+
+      archive = ChatChannelArchive.create!(
+        chat_channel: chat_channel,
+        archived_by: acting_user,
+        total_messages: chat_channel.chat_messages.count,
+        destination_topic_id: topic_params[:topic_id],
+        destination_topic_title: topic_params[:topic_title],
+        destination_category_id: topic_params[:category_id],
+        destination_tags: topic_params[:tags],
+      )
+      Jobs.enqueue(:chat_channel_archive, chat_channel_archive_id: archive.id)
+
+      archive
+    end
+  end
+
+  attr_reader :chat_channel_archive, :chat_channel
+
+  def initialize(chat_channel_archive)
+    @chat_channel_archive = chat_channel_archive
+    @chat_channel = chat_channel_archive.chat_channel
+  end
+
+  def execute
+    chat_channel_archive.update(archive_error: nil)
+
+    begin
+      ensure_destination_topic_exists!
+
+      Rails.logger.info("Creating posts from message batches for #{chat_channel.name} archive, #{chat_channel_archive.total_messages} messages to archive (#{chat_channel_archive.total_messages / ARCHIVED_MESSAGES_PER_POST} posts).")
+
+      # a batch should be idempotent, either the post is created and the
+      # messages are deleted or we roll back the whole thing.
+      #
+      # at some point we may want to reconsider disabling post validations,
+      # and add in things like dynamic resizing of the number of messages per
+      # post based on post length, but that can be done later
+      #
+      # another future improvement is to send a MessageBus message for each
+      # completed batch, so the UI can receive updates and show a progress
+      # bar or something similar
+      chat_channel.chat_messages.find_in_batches(
+        batch_size: ARCHIVED_MESSAGES_PER_POST
+      ) do |chat_messages|
+        create_post(
+          ChatTranscriptService.new(
+            chat_channel, messages: chat_messages
+          ).generate_markdown
+        ) do
+          delete_message_batch(chat_messages.map(&:id))
+        end
+      end
+
+      complete_archive
+    rescue => err
+      notify_archiver(:failed, error: err)
+      raise err
+    end
+  end
+
+  private
+
+  def create_post(raw)
+    pc = nil
+    Post.transaction do
+      pc = PostCreator.new(
+        chat_channel_archive.archived_by,
+        raw: raw,
+
+        # we must skip these because the posts are created in a big transaction,
+        # we do them all at the end instead
+        skip_jobs: true,
+
+        # we do not want to be sending out notifications etc. from this
+        # automatic background process
+        import_mode: true,
+
+        # don't want to be stopped by watched word or post length validations
+        skip_validations: true,
+
+        topic_id: chat_channel_archive.destination_topic_id
+      )
+
+      pc.create
```

Comment:
I am not sure, but when we create a new post for archive, do we want to relink other models like user_actions, uploads etc?

Something which is done by PostMover https://github.com/discourse/discourse/blob/main/app/models/post_mover.rb#L81;L88

Reply:
If we decided to move those items, then it should be done later in `delete_message_batch`

Reply:
I don't think there is anything to relink really...I already checked that uploads are linked to the post correctly in this spec:

https://github.com/discourse-org/discourse-topic-chat/blob/b8266e0ad02c297b8e2c7c8138656f6179854542/spec/lib/chat_channel_archive_service_spec.rb#L112-L118

Reply:
Thank you, I understand it now :)